### PR TITLE
Add max_submaps_with_pure_localization parameter

### LINKED ISSUE #1116: kSubmapsToKeep config parameter for pure localization
A config parameter for how many submaps are kept during pure localization would be good.
Currently it is hard coded [here](https://github.com/googlecartographer/cartographer/blob/41f17e57cd6c3af382399db15507f70c5f7865fc/cartographer/mapping/map_builder.cc#L139) to `3`:

```c++
  if (trajectory_options.pure_localization()) {
    constexpr int kSubmapsToKeep = 3;
    pose_graph_->AddTrimmer(common::make_unique<PureLocalizationTrimmer>(
        trajectory_id, kSubmapsToKeep));
}
```

This can limit the pose optimization, especially if you run with a low `TRAJECTORY_BUILDER_3D.submaps.num_range_data`

### PR DESCRIPTION
**Pull request by AlexReimann**
Adds a parameter for the currently hard coded `kSubmapsToKeep` in `pure_localization` mode.

The new parameter `max_submaps_with_pure_localization` allows to specify the history size of submaps in `pure_localization` mode.

Closes #1116

### CODE REVIEW

**Comment by cschuet** [comment]
@AlexReimann Thanks for the PR! If we start adding options to trimmers, I think it is cleaner to follow a similar approach as we did for the "overlapping_submaps_trimmer_2d", i.e. an optional dictionary-valued key "pure_localization_trimmer" and the options below that, i.e.

```
TRAJECTORY_BUILDER = {
  trajectory_builder_2d = TRAJECTORY_BUILDER_2D,
  trajectory_builder_3d = TRAJECTORY_BUILDER_3D,
  pure_localization_trimmer = {
    max_submaps_to_keep = 3,
  },
  overlapping_submaps_trimmer_2d = {
    fresh_submaps_count = 1,
    min_covered_area = 2,
    min_added_submaps_count = 5,
  },
}
```

@pifon2a WDYT?

**Comment by pifon2a** [comment]
@cschuet Agreed.

**Comment by cschuet** [comment]
@pifon2a Can you take a look at this again, merge it if it looks good and send follow-up PRs to other repos as needed? Thanks.

File: configuration_files/trajectory_builder.lua [hunk 1]
```diff
@@ -19,4 +19,7 @@ TRAJECTORY_BUILDER = {
   trajectory_builder_2d = TRAJECTORY_BUILDER_2D,
   trajectory_builder_3d = TRAJECTORY_BUILDER_3D,
```
Comment (pifon2a) [inline configuration_files/trajectory_builder.lua]: can we get rid of `pure_localization` flag in lua and proto as well? having `pure_localization_trimmer` should be enough

Comment (AlexReimann) [reply configuration_files/trajectory_builder.lua]: So check if there is a `pure_localization_trimmer` and depending on that go into `pure_localization` mode?
Do you want that in this PR?

Comment (pifon2a) [reply configuration_files/trajectory_builder.lua]: yes

Comment (AlexReimann) [reply configuration_files/trajectory_builder.lua]: Finally got around to do the changes.

Please check.

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by pifon2a** [comment]
Please, also update the current lua files that use this option.

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 1.

**Comment by wally-the-cartographer** [comment]
Refusing to merge, since the combined checks state is failure

**Comment by gaschler** [comment]
It seems like the grpc handler is not updated yet:
`[94/410] Building CXX object CMakeFiles/cartographer.dir/cartographer/cloud/internal/handlers/add_trajectory_handler.cc.o
FAILED: /usr/bin/c++   -DUSE_PROMETHEUS=1 -O3 -DNDEBUG -isystem /usr/include/eigen3 -isystem /usr/local/include -isystem /usr/include/lua5.2 -isystem /usr/include/cairo -isystem /usr/include/glib-2.0 -isystem /usr/lib/x86_64-linux-gnu/glib-2.0/include -isystem /usr/include/pixman-1 -isystem /usr/include/freetype2 -isystem /usr/include/libpng12 -I. -I..     -pthread -std=c++11 -fPIC  -Wall -Wpedantic -Werror=format-security -Werror=missing-braces -Werror=reorder -Werror=return-type -Werror=switch -Werror=uninitialized -O3 -DNDEBUG -std=c++11 -MMD -MT CMakeFiles/cartographer.dir/cartographer/cloud/internal/handlers/add_trajectory_handler.cc.o -MF "CMakeFiles/cartographer.dir/cartographer/cloud/internal/handlers/add_trajectory_handler.cc.o.d" -o CMakeFiles/cartographer.dir/cartographer/cloud/internal/handlers/add_trajectory_handler.cc.o -c ../cartographer/cloud/internal/handlers/add_trajectory_handler.cc
../cartographer/cloud/internal/handlers/add_trajectory_handler.cc: In member function 'virtual void cartographer::cloud::handlers::AddTrajectoryHandler::OnRequest(const cartographer::cloud::proto::AddTrajectoryRequest&)':
../cartographer/cloud/internal/handlers/add_trajectory_handler.cc:56:32: error: 'class cartographer::mapping::proto::TrajectoryBuilderOptions' has no member named 'set_pure_localization'
     trajectory_builder_options.set_pure_localization(false);`

**Comment by AlexReimann** [comment]
I'm wondering about the freezing part of [this comment in the cloud trajectory handler](https://github.com/googlecartographer/cartographer/blob/b5279532ce9e2a511f688fdf3bf06c4331a027e7/cartographer/cloud/internal/handlers/add_trajectory_handler.cc#L54):

```c++
    // Don't instantiate the 'PureLocalizationTrimmer' on the server and don't
    // freeze the trajectory on the server.
```

I didn't change anything about the freezing, is something missing?

Also I just commented the `pure_localization_trimmer` out in the lua to keep it an example.
Should this be fully removed instead?

This PR will also need some changes in the configuration files of the ROS repos

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 2.

File: cartographer/mapping/proto/trajectory_builder_options.proto [hunk 2]
```diff
@@ -29,7 +29,6 @@ message InitialTrajectoryPose {
 message TrajectoryBuilderOptions {
   LocalTrajectoryBuilderOptions2D trajectory_builder_2d_options = 1;
   LocalTrajectoryBuilderOptions3D trajectory_builder_3d_options = 2;
-  bool pure_localization = 3;
```
Comment (gaschler) [inline cartographer/mapping/proto/trajectory_builder_options.proto]: This breaks existing pbstreams. Is it worth it?

Comment (pifon2a) [reply cartographer/mapping/proto/trajectory_builder_options.proto]: We can make it backwards-compatible. I'll send the PR.

**Pull request merged**